repos/os/src/drivers/platform/spec/x86/pci_config_access.h

```diff
@@ -100,26 +89,34 @@ namespace Platform {
 			              unsigned char addr, Device::Access_size size,
 			              bool track = true)
 			{
-				/* write target address */
-				_io_port<REG_ADDR>()->outl(REG_ADDR, _cfg_addr(bus, device, function, addr));
+				unsigned ret;
+				const addr_t mmconf = _base + _dev_base(bus, device, function);
+				Attached_io_mem_dataspace mmio(_env, mmconf, MMCONF_REGION_SIZE);
+				char * field = mmio.local_addr<char>() + addr;
 
-				/* return read value */
+				/*
+				 * Explicitly use EAX register to make this work for Muen and
+				 * general x86. Furthermore, clear excess bits before return to
+				 * guarantee requested size.
+				 */
 				switch (size) {
 				case Device::ACCESS_8BIT:
 					if (track)
 						_use_register(addr, 1);
-
-					return _io_port<REG_DATA>()->inb(REG_DATA + (addr & 3));
+					asm volatile("movb %1,%%al" :"=a" (ret) :"m" (*((volatile unsigned char *)field)) :"memory");
```

Comment:
IMO dereferencing a `unsigned char *` pointer should suffice here. Do you agree?

Reply:
Instead of `volatile unsigned char *`?

Will drop superfluous parenthesis in the `ACCESS_8BIT` case.

Reply:
Sorry, my note was somewhat misleading. What I actually question is the use of inline assembler for a simple `unsigned char ret = *(unsigned char volatile *)field`. Is there any need for the assembler code?

Reply:
You are totally right, for the non-Muen case a dereference would probably suffice. On Muen however, our simplified device emulation code always returns `0xffff` in RAX to indicate a non-existing device (see https://git.codelabs.ch/?p=muen.git;a=blob;f=components/sm/src/devices-pciconf.adb;h=a953b38c957214a2865178b759b9e14c53687a8b;hb=dd59337163c174d584b5e1245b9278e2a109e5f8#l662). 

This simple method works well with Linux guests. 

Also, the inline assembly is inspired by the `read*` functions of the Linux kernel (`arch/x86/include/asm/io.h`) which also uses assembly on x86.